src/CommunityToolkit/Xamarin.CommunityToolkit/Views/Calendar/Calendar.shared.cs

```diff
@@ -0,0 +1,549 @@
+
+					foreach (var calendarDay in daysToRemove)
+					{
+						gridDays.Children.Remove(calendarDay);
+					}
+				}
+			}
+		}
+
+		void UpdateDays()
+		{
+			var weeksInMonth = Date.WeeksInMonth(FirstDayOfWeek);
+			var daysInMonth = Date.DaysInMonth();
+
+			for (var day = 1; day <= daysInMonth; day++)
+			{
+				var date = new DateTimeOffset(Date.Year, Date.Month, day, Date.Hour, Date.Minute, Date.Second, Date.Offset);
+				var weekOfMonth = date.WeekOfMonth(FirstDayOfWeek);
+
+				UpdateDaysFromPreviousOrNextMonths(weekOfMonth, date, weeksInMonth, daysInMonth);
+
+				if (!ShowWeekends && (date.DayOfWeek is DayOfWeek.Saturday or DayOfWeek.Sunday))
+				{
+					continue;
+				}
+
+				UpdateDay(date, weekOfMonth, isVisible: true);
+			}
+
+			if (SelectedDays?.Count > 0)
+			{
+				SelectDays(SelectedDays);
+			}
+		}
+
+		void UpdateDaysFromPreviousOrNextMonths(
+			int week,
+			DateTimeOffset date,
+			int weeksInMonth,
+			int daysInMonth)
+		{
+			var lastDayOfWeek = FirstDayOfWeek.PreviousOrFirst();
+			
+			if (week is 1
+				&& date.Day is 1
+				&& date.DayOfWeek != FirstDayOfWeek)
+			{
+				var newDate = date;
+
+				do
+				{
+					newDate = newDate.AddDays(-1);
+
+					if (!ShowWeekends && (newDate.DayOfWeek is DayOfWeek.Saturday or DayOfWeek.Sunday))
+					{
+						continue;
+					}
+
+					UpdateDay(newDate, week, isVisible: ShowDaysFromOtherMonths);
+				} while (newDate.DayOfWeek != FirstDayOfWeek);
+			}
+			else if (week == weeksInMonth
+					 && date.Day == daysInMonth
+					 && date.DayOfWeek != lastDayOfWeek)
+			{
+				var newDate = date;
+
+				do
+				{
+					newDate = newDate.AddDays(1);
+
+					if (!ShowWeekends && (newDate.DayOfWeek is DayOfWeek.Saturday or DayOfWeek.Sunday))
+					{
+						continue;
+					}
+
+					UpdateDay(newDate, week, isVisible: ShowDaysFromOtherMonths);
+				} while (newDate.DayOfWeek != lastDayOfWeek);
+			}
+		}
+
+		void UpdateDay(DateTimeOffset date, int week, bool isVisible)
+		{
+			var column = date.DayOfWeek(FirstDayOfWeek, includeWeekends: ShowWeekends) - 1;
+			var row = week - 1;
+
+			var calendarDay = days.FirstOrDefault(x => Grid.GetColumn(x) == column && Grid.GetRow(x) == row);
+
+			if (calendarDay == null)
+			{
+				calendarDay = new CalendarDay
+				{
+					HorizontalOptions = LayoutOptions.CenterAndExpand,
+					VerticalOptions = LayoutOptions.CenterAndExpand,
+					ControlTemplate = DayControlTemplate,
+					Date = date,
+				};
+
+				Grid.SetColumn(calendarDay, column);
+				Grid.SetRow(calendarDay, row);
+
+				var tapGestureRecognizer = new TapGestureRecognizer();
+				tapGestureRecognizer.Tapped += OnCalendarDayTapped;
+				calendarDay.GestureRecognizers.Add(tapGestureRecognizer);
+
+				days.Add(calendarDay);
+				gridDays.Children.Add(calendarDay);
+			}
+			else
+			{
+				calendarDay.Date = date;
+				calendarDay.ControlTemplate = DayControlTemplate;
+			}
+
+			calendarDay.Opacity = isVisible ? 1 : 0;
```

Comment:
@brminnick this line was originally wrapped into `Device.BeginInvokeOnMainThread` because I've noticed some visible delay between when day is updated with new value. Below is video that shows this problem:

https://user-images.githubusercontent.com/12022141/120119433-f0da6880-c197-11eb-9527-a9e207742b7f.mov

When switching to next month, days in the first row have old value for a slight of second before correct date is visible. Wrapping this line in `Device.BeginInvokeOnMainThread` solves this issue.

Let me know what you think.

Reply:
Looks like that behavior only happens when `Show days from other months` is not enabled (video attached).

I'll investigate why. It's likely just an order-of-operations bug; i.e. We're changing the days before changing the month.

https://user-images.githubusercontent.com/13558917/120235760-9aa90c00-c20f-11eb-87c6-506945f157ec.mp4

Reply:
I will take a look at this

Reply:
@brminnick  writing code like this:

```
        else
	{
		calendarDay.Date = date;
	}

	if (isVisible)
	{
		OnDayUpdated(calendarDay);
	}

	if (Device.RuntimePlatform.Equals(Device.iOS))
	{
		await Task.Delay(1);
	}

	calendarDay.Opacity = isVisible ? 1 : 0;
}
```

Solves the issue on iOS. Let me know what you think

Reply:
What's the root cause, though?

`await Task.Delay` is the same workaround as `Device.BeginInvokeOnMainThread`; both generate a thread/context switch which negatively impacts performance, and I'd like to avoid that.

Reply:
I couldn't figure out the root cause :/ Even if we set `calendarDay.Date = date` before setting `calendarDay.Opacity = 1`, it's still being updated visually AFTER CalendarDay becomes visible, which causes that for a slight of second old value is visible. It seems like iOS is not updating visual presence of UI component until it is visible to user.

Reply:
@brminnick I've created simple project to reproduce this issue: https://github.com/DamianAntonowicz/XamarinFormsIosOpacityTest

Here is a video with issue reproduced:
https://user-images.githubusercontent.com/12022141/120925179-aeed7d00-c6d7-11eb-87d2-ff6d8bf5784a.mov

It is not happening every time, but it happens and is visible.

The code for this is very simple:
```
private void ButtonOpacityToZero_OnClicked(object sender, EventArgs e)
        {
            CalendarDay1.Opacity = 0;
 
            ...
        }
        
        private void ButtonOpacityToOne_OnClicked(object sender, EventArgs e)
        {
            var random = new Random();
            
            CalendarDay1.Date = DateTimeOffset.Now.AddDays(random.Next(0, 10));
            CalendarDay1.Opacity = 1;

            ...
        }```

Do you have any suggestions? :)

Reply:
@brminnick did you had time to take a look at this issue?

Reply:
@pictos could you take a look at this conversation? Maybe you will have some idea :)

Reply:
can this be a bug on XF?
> I only see this behavior happening on iOS; it is not happening on my Android simulator.

From Brandon's comment, this looks like a platform-specific bug, in other words, how does iOS handle the rendering. If I'm not wrong we can disable animations on iOS, maybe this can solve the issue. We disable the animation to change values and re-enabled it after setting `Opacity` to 1.

@DamianAntonowicz can you test it? I don't have iOS environment configured

Reply:
@pictos thanks for the tip! I've fixed the issue with following code:
```
calendarDay.FadeTo(isVisible ? 1 : 0, length: 0);
```

Reply:
@DamianAntonowicz This does not resolve the issue.

`FadeTo` returns a `Task`, which we have to `await`. 

This then requires us to update the following methods to make them asynchronous:
- `async Task UpdateDay`
- `async Task UpdateDaysFromPreviousOrNextMonths`
- `async Task UpdateDays`
- `async Task UpdateCalendarDays`
- `async Task UpdateCalendarDays`
- `static async void OnDayControlTemplateChanged`
- `static async void OnFirstDayOfWeekChanged`
- `static async void OnShowWeekendsChanged`
- `static async void OnShowDaysFromOtherMonthsPropertyChanged`

After updating those methods, the Calendar UI on both iOS + Android then has a ripple effect where the days appear & disappear in sequence instead of in parallel:
![Simulator Screen Recording - iPhone 12 - 2021-07-28 at 16 28 29](https://user-images.githubusercontent.com/13558917/127399479-8ba510d5-d5c0-4476-81da-aa5c652c1b5f.gif)

Reply:
I see two paths for fixing this bug:
- Keep `await FadeTo`, and update your logic to batch the day changes so that the days update in parallel
- Revert [this commit](https://github.com/xamarin/XamarinCommunityToolkit/pull/1331/commits/ee9788fdde518a01e63b9ffe195d6d43a122ef1a) and fix the original problem on iOS. 
  - it may require a custom renderer on iOS

Reply:
@brminnick I've rushed with this change, sorry for that. Could you point me in proper direction with custom renderer on iOS? I'm not sure how this renderer should be created.

Reply:
@brminnick @pictos I've updated Calendar logic to update days in parallel. I think it fixed update issue on iOS - at least I cannot reproduce issue anymore. 

Let me know what you think about code changes.